src/cli/test/reporters/github/index.ts

```diff
@@ -0,0 +1,66 @@
+import { Context, Reporter } from '@jest/reporters';
+import { AggregatedResult, TestResult } from '@jest/test-result';
+import pMap from 'p-map';
+
+import * as GitHub from '../../../../api/github';
+import {
+  buildNameFromEnvironment,
+  enabledFromEnvironment,
+} from '../../../../api/github/environment';
+
+import { createAnnotations } from './annotations';
+
+const DEFAULT_NAME = 'noDisplayName';
+
+export default class GitHubReporter implements Pick<Reporter, 'onRunComplete'> {
+  async onRunComplete(
+    _contexts: Set<Context>,
+    results: AggregatedResult,
+  ): Promise<void> {
+    if (!enabledFromEnvironment()) {
+      return;
+    }
+
+    // Sort tests by display name
+    const sortedTestResults: Record<string, TestResult[]> = {};
+    results.testResults.forEach((testResult) => {
+      const displayName = testResult.displayName?.name ?? DEFAULT_NAME;
+      sortedTestResults[displayName] ??= [];
+      sortedTestResults[displayName].push(testResult);
+    });
+
+    // Create annotations for each display name
+    const annotationResults = Object.entries(sortedTestResults).map<{
+      displayName: string;
+      annotations: GitHub.Annotation[];
+    }>(([displayName, testResults]) => ({
+      displayName,
+      annotations: createAnnotations(testResults),
+    }));
+
+    // Create a check run per display name. Run in series.
+    await pMap(
```

Comment:
If we're running these serially, I think we'd be better off with a plain old `for...of` loop.